---
title: "Mage Overview"
author: "Charles Bernard"
date: "2018"
output: 
  rmarkdown::html_vignette:
    fig_caption: yes
    keep_md: yes
    number_sections: yes
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
devtools::load_all();
```

# Prerequisites

The `mage` package has been designed to capture but foremost to classify any significant association 
between the expression of 2 genes in a single cell rna sequencing dataset.

## What does it mean when 2 genes are said to be associated?

The best way to figure yourself out what an association between 2 genes really means would be probably to 
visualize it as a scatterplot with the x and y axis corresponding to their respective expression,
and in which each point would represent a cell. 

As a matter of example, let us produce three scatterplots:

- a linear correlation
- a non-coexistence relationship
- a random relationship

```{r three_scatterplots, fig.show='hold'}
plot(1:10, xlab = "Gene A", ylab = "Gene B", main = "Linear Correlation");
plot(c(0:4, rep(0,5)), c(rep(0,5), 0:4), xlab = "Gene C", ylab = "Gene D", main = "Non-Coexistence");
plot(runif(10, 0, 10), runif(10, 0, 10), xlab = "Gene E", ylab = "Gene F", main = "Random");
```

Intuitevely, we can quickly see that the two first scatterplots result from a meaningful relationship 
between 2 genes while the third one depict a weaker association. With this respect, the two first gene
pairs must be considered as "significantly associated". 

When 2 genes are said to be "significantly associated", this somehow means that their relationship 
is "remarkable" and clearly distinguishable from noise.

The MINE statistics have been developped first, to estimate the strength of an association between 2 variables
no matter what type of relationship may link them together and second, to characterize any given association
by testing some propreties of the shape of the relationship *(non-linearity, complexity, non-monotonicity ...)*.

The `mage` package rely on these statistics to cluster the associations by type of relationships.

## What is the output of mage?

... to be written

# Tutorial Presentation

This tutorial provides a quick overview of the different features this package can offer. 
It shows how to compute the association strength for each unique pair of genes that 
can be composed from a gene expression matrix, how to characterize them and more essentially 
how to make the most of the variety of the `mage` functions to cluster the significant associations 
based on the type of their relationships *(i.e linear correlation, non-coexistence, sinusoids ...)*.

In a near future, this tutorial will include a series of post-hoc analyses to explain how the standard
output from mage can be used to extract meaningful informations from a biological point of view ...

# Setup

The first step is to load the mage R package as well as the toy dataset that will be used for this tutorial.
The dataset is a gene expression matrix (raw count) used by the Seurat R package as canonical input for
the code examples. It is composed of 80 cells for 230 genes. 

```{r tuto_setup, message=FALSE, warning=TRUE}
# Load the mage library
library(mage);

# Load the gene expression matrix (raw count) that will be
# used for this tutorial. 
data(pbmc_small_raw_data);
gene_exp_mat <- pbmc_small_raw_data;

# Show all the informations about this dataset:
help(pbmc_small_raw_data);
```

# Compute the set of MINE scores for all unique gene pairs

Our toy dataset contains 230 genes, which result exactly in $230(230-1)/2 = 26335$ unique gene pairs

```{r compute_scores}
# Notice that the scores are computed in parallel. 
scores <- compute_scores(gene_exp_mat,
                         n.cores = parallel::detectCores() - 1);

# MEV happens to be redundant with MIC so we can actually get rid of it
scores <- scores[, -"MEV (functionality)"];
```

This is what the table of scores look like for the top 5 associations

```{r show_scores_tab}
knitr::kable(scores[1:5, ]);
```

# input for the Seurat R package examples.


Vignettes are long form documentation commonly included in packages. Because they are part of the distribution of the package, they need to be as compact as possible. The `html_vignette` output type provides a custom style sheet (and tweaks some options) to ensure that the resulting html is as small as possible. The `html_vignette` format:

- Never uses retina figures
- Has a smaller default figure size
- Uses a custom CSS stylesheet instead of the default Twitter Bootstrap style

## Vignette Info

Note the various macros within the `vignette` section of the metadata block above. These are required in order to instruct R how to build the vignette. Note that you should change the `title` field and the `\VignetteIndexEntry` to match the title of your vignette.

## Styles

The `html_vignette` template includes a basic CSS theme. To override this theme you can specify your own CSS in the document metadata as follows:

    output: 
      rmarkdown::html_vignette:
        css: mystyles.css

## Figures

The figure sizes have been customised so that you can easily put two images side-by-side. 

```{r, fig.show='hold'}
plot(1:10)
plot(10:1)
``` 

You can enable figure captions by `fig_caption: yes` in YAML:

    output:
      rmarkdown::html_vignette:
        fig_caption: yes

Then you can use the chunk option `fig.cap = "Your figure caption."` in **knitr**.

## More Examples

You can write math expressions, e.g. $Y = X\beta + \epsilon$, footnotes^[A footnote here.], and tables, e.g. using `knitr::kable()`.

```{r, echo=FALSE, results='asis'}
knitr::kable(head(mtcars, 10))
```

Also a quote using `>`:

> "He who gives up [code] safety for [code] speed deserves neither."
([via](https://twitter.com/hadleywickham/status/504368538874703872))
